src/behavioural-events/flush-manager.ts

```diff
@@ -0,0 +1,95 @@
+import { Logger } from "../helpers/logger";
+
+export class FlushManager {
+  private readonly initialDelay: number;
+  private readonly maxDelay: number;
+  private readonly callback: () => Promise<void>;
+  private currentDelay: number;
+  private timeoutId: ReturnType<typeof setTimeout> | undefined = undefined;
+  private executingCallback: boolean = false;
+
+  constructor(
+    initialDelay: number,
+    maxDelay: number,
+    callback: () => Promise<void>,
+  ) {
+    this.initialDelay = initialDelay;
+    this.currentDelay = initialDelay;
+    this.maxDelay = maxDelay;
+    this.callback = callback;
+  }
+
+  public tryFlush() {
+    if (this.backingOff()) {
+      Logger.log(`Backing off, not flushing`);
+      return;
+    }
+
+    Logger.log(`Flushing immediately`);
+    this.clearTimeout();
+    this.executeCallbackWithRetries();
+  }
+
+  public stop() {
+    this.clearTimeout();
+  }
+
+  public schedule(delay?: number) {
+    if (this.timeoutId !== undefined) {
+      Logger.log(`Already scheduled`);
+      return;
+    }
+
+    Logger.log(`Scheduling callback after ${this.currentDelay}ms delay`);
```

Comment:
Should we use `debugLog` instead? (Applies to several lines in this file, and maybe other)

Reply:
Yes! 🤦🏻  going to review the rest in case there are any other.